### ISSUE TITLE
Creation of custom [derived] field types

### ISSUE THREAD

**Issue by julien-f** [natevw/struct-fu#7]
It would be nice if it was possible to easily create custom field types from an existing one.

Here is a trivial (and probably incorrect) example for implementing a 64 bits unsigned integer:

```js
const COEFF = Math.pow(2, 32)

const uint64 = _.derive(_.uint32(2), {
  pack (originalPack, val, offset) {
    return originalPack([ Math.floor(val / COEFF), val % COEFF ], offset)
  },
  unpack (originalUnpack, buf, offset) {
    const val = originalUnpack(buf, offset)
    return val[0] * COEFF + val[1]
  }
})
```

What's your opinion on this?

**Comment by natevw** [comment]
Interesting idea.

I think supporting custom field types is important; my first reaction was that this wasn't generic enough. But fully polishing/exposing/documenting the internal API for external use is a bigger project and I think this could cover a lot of use cases in the meantime. And we might always want something like this `derive` even with a "fuller" solution.

I might tweak it just a bit. I'll stick with your example since it has clear intent, even if JS only supports [~53-bit integers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) ;-)

```
const uint64_array = _.derive('somename', _.uint32(2), function myPacker(val) {
    return [ Math.floor(val / COEFF), val % COEFF ];
}, function myUnpacker(val) {
    return val[0] * COEFF + val[1]
}, 42)
```

Specifically the changes I've made are:

* clarify the that `derive` field can optionally, like all other fields, take a name and count. (We could maybe use the name of the derived field instead, but we want to allow a separate count and so IMO just keep name separate too.)
* there's a bit of dilemma in that `pack` and `unpack` are aliases for `bytesFromValue` and `valueFromBytes` and its the latter which are used internally. Maybe it's cleaner to require two functions instead of an object? [More on this in followup comment…]
* finally, instead of needing a `super` function — and with it exposing the `offset` stuff! — these could be simple value-transforming functions.

How's that seem to you?

**Comment by julien-f** [comment]
What I meant was not to create an instance of a type but a type itself, that's why I did not add a name or a count.

Apart from that, you're propositions look fine to me :)

**Comment by natevw** [comment]
@julien-f Ah, that's an important clarification and I agree that's a better design. So it would look like:

```
let uint64_t = _.derive(_.uint32(2), function myPacker(val) {
    return [ Math.floor(val / COEFF), val % COEFF ];
}, function myUnpacker(val) {
    return val[0] * COEFF + val[1]
});

let mystruct = _.struct([
    _.char('header', 8),
    uint64_t('lots_of_big_numbers', 42)
]);
```

Since `_.derive` would always take exactly 3 arguments (rather than optional names/counts like I got mixed up with) the potential anonymous functions bother me even less.

**Comment by julien-f** [comment]
Yep, looks nice.

The idea is to be able to implement types outside of the core, for instance, @nraynaud PR would have been less necessary if it was possible to do the same in userland.

**Comment by natevw** [comment]
Finally found some time to add this, ended up being fairly simple but caught a few issues already via some tests. If you're still interested, you can give it a try via `npm install natevw/struct-fu#derived`.

I'll wait a bit for feedback and do a little sleeping on it, otherwise I think it should be able to go out as a v1.2.

**Comment by natevw** [comment]
[Whoops, has tests and implementation but still needs documentation, which could possibly drive a bit more tweaks yet… I've noted that on my pending branch merge: https://github.com/natevw/struct-fu/pull/12]

**Comment by natevw** [comment]
Sorry, I thought this had been finished… I wrote some documentation today and just published to npm as `struct-fu@1.2.0`.

**Comment by julien-f** [comment]
No problem, thank you for this :)
